When you select a star, indicate that the add-on rating has been saved

bobsilverberg opened the pull request:
Fixes #5941 

I ended up having to duplicate some code between `RatingManager` and `AddonReviewManager` because the latter is not displayed when there is no review body present, and it's where the flash message is currently rendered. There may be a refactoring that will eliminate this duplication, but the way the code is currently structured it's not obvious to me.

![sep-11-2018 12-39-05](https://user-images.githubusercontent.com/142755/45374323-bb445a00-b5bf-11e8-9aff-2f88a74f0d41.gif)

![sep-11-2018 12-48-22](https://user-images.githubusercontent.com/142755/45374748-06ab3800-b5c1-11e8-9c11-c25d3f5ce8b4.gif)

kumar303 commented on src/amo/components/RatingManager/index.js:
```diff
@@ -251,6 +288,25 @@ export class RatingManagerBase extends React.Component<InternalProps, State> {
+          >
+            {flashMessage === STARTED_SAVE_RATING
+              ? i18n.gettext('Saving')
+              : i18n.gettext('Saved')}
```
I like how the mock spells it out exactly as *Star rating saved* so I think we should do that.

<img width="386" alt="screenshot 2018-09-11 12 07 59" src="https://user-images.githubusercontent.com/55398/45375697-64d51c80-b5bb-11e8-9582-3d121d8bc920.png">

The mock also calls for a checkbox icon but, you know, I think it would be nicer to just render a proper notice for this (which gets you a checkbox icon):

```js
<Notice type="success">{i18n.gettext('Saved...')}</Notice>
```

What do you think? The other UX mock already has this notice type for saving a review. I don't see why we can't use it in both places:

<img width="385" alt="screenshot 2018-09-11 12 10 11" src="https://user-images.githubusercontent.com/55398/45375810-b1b8f300-b5bb-11e8-91ee-6d51958c8d2f.png">

kumar303 commented on src/amo/components/RatingManager/styles.scss:
```diff
@@ -60,3 +60,16 @@
   width: 100%;
 }
+
+.RatingManager-savedRating {
```
Could we pull this logic into a new component like `RatingManagerNotice` so that it can be shared?

bobsilverberg commented:
I've updated this and introduced the `RatingManagerNotice`. At this point it doesn't really buy much, as I've really just moved code from `RatingManager` into `RatingManagerNotice` but we should be able to use it more in the future.

Here are a couple of new screencasts:

![sep-11-2018 15-27-55](https://user-images.githubusercontent.com/142755/45382858-952ab400-b5d7-11e8-87bf-ff45cb2e1189.gif)

![sep-11-2018 15-26-51](https://user-images.githubusercontent.com/142755/45382865-9bb92b80-b5d7-11e8-9d0c-b8861d1767e6.gif)

kumar303 commented on src/amo/components/RatingManager/index.js:
```diff
@@ -28,10 +34,12 @@ import {
 import log from 'core/logger';
 import Button from 'ui/components/Button';
+import { genericType, successType } from 'ui/components/Notice';
+import RatingManagerNotice from 'ui/components/RatingManagerNotice';
```
This seems like a pretty specific component so I'd expect it in `amo/components`. I guess it doesn't really matter, though.

bobsilverberg replied:
Sure, I can move it.

kumar303 commented on src/ui/components/RatingManagerNotice/index.js:
```diff
@@ -0,0 +1,35 @@
+  className?: string | null,
+  hideMessage: boolean,
+  message: string,
+  type: NoticeType,
```
If the caller omits `type` (or something like that), you could render it without `Notice` and use this component for the existing `<span>` notification. What do you think? It seems to work except I didn't add in the CSS transition.

<details>
<summary>Example patch</summary>

```diff
diff --git a/src/amo/components/AddonReviewManager/index.js b/src/amo/components/AddonReviewManager/index.js
index 4ec38f176..b9e8f45c5 100644
--- a/src/amo/components/AddonReviewManager/index.js
+++ b/src/amo/components/AddonReviewManager/index.js
@@ -15,6 +15,7 @@ import translate from 'core/i18n/translate';
 import { sanitizeHTML } from 'core/utils';
 import DismissibleTextForm from 'ui/components/DismissibleTextForm';
 import Rating from 'ui/components/Rating';
+import RatingManagerNotice from 'ui/components/RatingManagerNotice';
 import type { AppState } from 'amo/store';
 import type { DispatchFunc } from 'core/types/redux';
 import type { ErrorHandlerType } from 'core/errorHandler';
@@ -98,17 +99,18 @@ export class AddonReviewManagerBase extends React.Component<InternalProps> {
             styleSize="small"
             yellowStars
           />
-          <span
-            className={makeClassName('AddonReviewManager-savedRating', {
-              'AddonReviewManager-savedRating-hidden':
-                flashMessage !== STARTED_SAVE_RATING &&
-                flashMessage !== SAVED_RATING,
-            })}
-          >
-            {flashMessage === STARTED_SAVE_RATING
-              ? i18n.gettext('Saving')
-              : i18n.gettext('Saved')}
-          </span>
+          <RatingManagerNotice
+            className="AddonReviewManager-savedRating"
+            hideMessage={
+              flashMessage !== STARTED_SAVE_RATING &&
+              flashMessage !== SAVED_RATING
+            }
+            message={
+              flashMessage === STARTED_SAVE_RATING
+                ? i18n.gettext('Saving')
+                : i18n.gettext('Saved')
+            }
+          />
         </div>
         <DismissibleTextForm
           formFooter={formFooter}
diff --git a/src/ui/components/RatingManagerNotice/index.js b/src/ui/components/RatingManagerNotice/index.js
index b3bc404ec..4bdccef56 100644
--- a/src/ui/components/RatingManagerNotice/index.js
+++ b/src/ui/components/RatingManagerNotice/index.js
@@ -11,7 +11,7 @@ type Props = {|
   className?: string | null,
   hideMessage: boolean,
   message: string,
-  type: NoticeType,
+  type?: NoticeType,
 |};
 
 const RatingManagerNotice = ({
@@ -20,16 +20,21 @@ const RatingManagerNotice = ({
   message,
   type,
 }: Props) => {
-  return (
-    <Notice
-      type={type}
-      className={makeClassName('RatingManagerNotice-savedRating', className, {
-        'RatingManagerNotice-savedRating-hidden': hideMessage,
-      })}
-    >
-      {message}
-    </Notice>
-  );
+  const props = {
+    className: makeClassName('RatingManagerNotice-savedRating', className, {
+      'RatingManagerNotice-savedRating-hidden': hideMessage,
+    }),
+  };
+
+  if (type) {
+    return (
+      <Notice type={type} {...props}>
+        {message}
+      </Notice>
+    );
+  }
+
+  return <span {...props}>{message}</span>;
 };
 
 export default RatingManagerNotice;
```

</details>

kumar303 commented on tests/unit/amo/components/TestRatingManager.js:
```diff
@@ -605,6 +676,49 @@ describe(__filename, () => {
+
+      const root = renderInline({ addon, errorHandler, store, version });
+
+      root.instance().onSelectRating(rating);
```
Since you don't need to wait for the promise (hooray), you can do this without reaching into the implementation of `RatingManager`. You can just get it from the `UserRating` configuration:

```diff
diff --git a/tests/unit/amo/components/TestRatingManager.js b/tests/unit/amo/components/TestRatingManager.js
index 6fd03ebfe..c177d6f67 100644
--- a/tests/unit/amo/components/TestRatingManager.js
+++ b/tests/unit/amo/components/TestRatingManager.js
@@ -687,7 +687,9 @@ describe(__filename, () => {
 
       const root = renderInline({ addon, errorHandler, store, version });
 
-      root.instance().onSelectRating(rating);
+      const review = root.find(UserRating);
+      const onSelectRating = review.prop('onSelectRating');
+      onSelectRating(5);
 
       sinon.assert.calledWith(
         dispatchSpy,
```

kumar303 commented on tests/unit/amo/components/TestRatingManager.js:
```diff
@@ -605,6 +676,49 @@ describe(__filename, () => {
+
+      const root = renderInline({ errorHandler, store });
+
+      root.instance().onSelectRating(rating);
```
Same here

bobsilverberg commented:
Thanks for the review, @kumar303. This is ready for another look.

kumar303 commented on src/amo/components/RatingManagerNotice/styles.scss:
```diff
@@ -0,0 +1,10 @@
+@import '~amo/css/styles';
+
+.RatingManagerNotice-savedRating {
+  color: $grey-50;
```
For some reason this is overriding `$green-90` which should be the text color for success notices.

<img width="352" alt="screenshot 2018-09-12 16 01 59" src="https://user-images.githubusercontent.com/55398/45453030-33388000-b6a5-11e8-83e0-d2a276b2d6d3.png">

bobsilverberg replied:
I'm not seeing this. I.e., I see the green message as you've shown above. But that doesn't really matter because we no longer need to request this grey colour for the component, so I've removed that style.

kumar303 commented:
Did you file an issue for changing `AddonReviewManager` to use `RatingManagerNotification`? We should get that at top of the list.

bobsilverberg commented on src/amo/components/RatingManager/styles.scss:
```diff
@@ -20,6 +20,9 @@
 }
 
 .RatingManager {
+  display: grid;
```
Sure, I'll undo those recent changes to the styles so that it's like it was in the previous iteration.

bobsilverberg commented:
> Did you file an issue for changing `AddonReviewManager` to use `RatingManagerNotification`? We should get that at top of the list.

I opened https://github.com/mozilla/addons-frontend/issues/6236

bobsilverberg commented:
@kumar303 Rebase has been done.

bobsilverberg commented on src/amo/components/RatingManagerNotice/styles.scss:
```diff
@@ -0,0 +1,9 @@
+@import '~amo/css/styles';
+
+.RatingManagerNotice-savedRating {
```
Yes, that does seem to be the correct place to add the bottom margin. Done!

bobsilverberg commented on src/amo/components/RatingManager/index.js:
```diff
@@ -246,6 +284,25 @@ export class RatingManagerBase extends React.Component<InternalProps, State> {
+                : i18n.gettext('Star rating saved')
+            }
+            type={
+              flashMessage === STARTED_SAVE_RATING ? genericType : successType
```
Yes, I can reproduce this. Should we file a follow-up, or do you anticipate this being addressed with some other future work currently in the queue?
